Fix to work with latest rust and syntex

webmobster opened the pull request:
Hi, I updated this to work with a more recent copy of rust and the latest syntex, the changes should be fairly small and non contraversial apart from the fix to do with the NodeId which I don't know enough about how the code works to understand, so I would appreciate any input on that one.

webmobster commented:
Didn't look deeply enough into the code at all, didn't realize there was feature switches, I'll update to pull request when I have sorted out the new issues

webmobster commented:
Ok I think, I have everything fixed and up to date but can't 100% confirm, as from looking at the tests in rustlex they never would have worked with syntex as they lack a build.rs, but everything tests and builds in nightly, and I got the "tests" in rustlex-codegen to work with both.
There is some ugliness in the new error handling, and would be willing to take suggestions on how it should be reworked with the the new way parse errors are emmited by syntex/rustc.

LeoTestard commented on codegen/src/codegen.in.rs:
```diff
@@ -111,7 +113,9 @@ pub fn lexer_struct(cx: &mut ExtCtxt, sp: Span, ident:Ident, props: &[Prop]) ->
-        P(ast::StructDef { ctor_id: None, fields: fields }),
+        //Copied this number from a rust ice message where an assertion failed, would appreciate
+        //input on wtf this is.
+        ast::VariantData::Struct(fields,4294967295),
```
Huh... Unfortunately, I don't have any idea either of what's going on here. We should look at the code of libsyntax, I suppose.

LeoTestard commented on Cargo.toml:
```diff
@@ -9,9 +9,16 @@ readme = "README.md"
+default     = []
+with-syntex = ["rustlex_codegen/with-syntex", "syntex"]
+
+
```
Mhh... If I recall correctly, the `rustlex_codegen` crate alone was sufficient to be used with syntex (look at the tests in codegen). The `rustlex` crate itself was just the compiler plugin and was used just as an entry point and didn't need to be compiled for stable at all. Did that change somehow? Is it needed with syntex too now?

LeoTestard commented on codegen/src/parser.rs:
```diff
@@ -92,33 +94,50 @@ impl<'a> Tokenizer for Parser<'a> {
+{
+        error.emit();
+        FatalError
+}
```
This is very unfortunate... It makes the code below much less readable. But I guess we have no choice. Maybe making it a trait would allow to call it with dot-notation and a shorter name, which would be a bit more readable I think. (Also, please, try to use the same coding style as the rest of the code for the braces and the indentation. :) )

LeoTestard commented on codegen/tests/Cargo.toml:
```diff
@@ -6,7 +6,7 @@ build = "build.rs"
 default     = ["nightly"]
 nightly     = ["rustlex"]
-with-syntex = ["rustlex_codegen/with-syntex", "syntex"]
+with-syntex = ["rustlex_codegen/with-syntex","rustlex/with-syntex", "syntex"]
```
Ah... didn't see this. Why do we need the `rustlex` crate here?

LeoTestard commented on codegen/tests/src/test.in.rs:
```diff
@@ -10,7 +10,9 @@ pub enum Token {
 
 rustlex! SimpleLexer {
     let A = 'a';
+    . => |lexer:&mut SimpleLexer<R>| None
```
Ooops. :D

LeoTestard commented:
Wow, thanks a lot for taking the time to do this long PR! And sorry for the late answer. Right now I don't have much time to review it in detail, so I just discussed a couple of things. If you can wait some time, I'll probably be able to merge it this week end. In the meantime, maybe @kali can help?

webmobster commented:
Hi, Thanks for looking at the pull request, hold of on merging it for now, as rust nightly (and therefore syntex) updated again shortly after my pull request and I havent had time to finish updating the code to support it again. I'll probably have some time on the weekend to do that and the changes should be smaller this time as the changes to the api is smaller.

LeoTestard commented:
Ok! Let me know when you have updated the PR then. Thanks a lot for investing time in RustLex!

webmobster commented:
Ok I have updated rustlex to the latest version of syntex (nightly and syntex_syntax are out of sync :/ so tests will never pass). Also as you noted in an earlier comment my hacky error conversion methods are not very nice and I completely agree, but the code would need a lot of rework to fix it properly. 
I would recommend that the code be fixed to a version of syntex and to drop pure nightly support, this would fix problems with the latest version of syntex being out of sync with rust nightlies and allow time to re-architect the code to properly work with the chosen version of syntex rather than this current race to unbreak things every 2 weeks. I have added this a separate pull request that "inherits" from this one.
